all-sdk/autotuning.md

```diff
@@ -0,0 +1,431 @@
+}
+trait WorkflowTaskSlotSupplier: SlotSupplier<SlotKind=WorkflowSlotKind> {}
+trait ActivityTaskSlotSupplier: SlotSupplier<SlotKind=ActivitySlotKind> {}
+trait LocalActivityTaskSlotSupplier: SlotSupplier<SlotKind=LocalActivitySlotKind> {}
+
+/// Users might want to be able to pause the handing-out of slots as an effective way of pausing their workers.
+/// We can provide an implementation for this that wraps their implementation, or one of the defaults we provide.
+struct PauseableSlotSupplier<T: SlotKind> {
+  inner: Arc<dyn SlotSupplier<SlotKind=T>>,
+  paused: AtomicBool,
+  // ... details
+}
+impl<T: SlotKind> PauseableSlotSupplier<T> {
+  pub fn new(supplier: impl SlotSupplier<SlotKind=T>) -> Self { /* ... */ }
+  pub fn pause(&self) { /* ... */ }
+  pub fn resume(&self) { /* ... */ }
+}
+impl<T: SlotKind> SlotSupplier for PauseableSlotSupplier<T> {
+  // ... implementation which checks the paused flag before handing out a slot
+}
+```
+
+## Workflow cache management
+
+The workflow cache is maintained independently of the number of slots in use. There could be 0 slots in use and 100
+workflows cached, and that's a totally normal scenario. We might also choose to make this user-configurable, though we
+certainly don't need to have that right out of the gate.
+
+Keeping our fixed-size implementation to start is fine, because we won't ask the user's slot provider for a slot unless
+there is room in the cache (either additional capacity, or an idle cached workflow that can be evicted). This avoids
+a potential problem where we ask for a slot, they provide one, but there's nothing we can evict at the moment.
+
+If and when we do introduce user-configurable caching, it's likely that they would want the same thing to be responsible
+for both slots and cache management, since the two are closely related.
+
+A possible future:
+
+```rust
+trait WorkflowCacheSizer {
+  /// Return true if it is acceptable to cache a new workflow. Information about already-in-use slots, and just-received
+  /// task is provided. Will not be called for an already-cached workflow who is receiving a new task.
+  /// 
+  /// Because the number of available slots must be <= the number of workflows cached, if this returns false
+  /// when there are no idle workflows in the cache (IE: All other outstanding slots are in use), we will buffer the
+  /// task and wait for another to complete so we can evict it and make room for the new one.
+  fn can_allow_workflow(&self, slots_info: &WorkflowSlotsInfo, new_task: &WorkflowSlotInfo) -> bool;
+  /// Called when deciding what workflow should be evicted from the cache. May return the run id of a workflow to evict.
+  /// If None, or an invalid run id is returned, the default LRU eviction / strategy is used. The id returned must
+  /// be one of the workflows provided in `evictable` (the ones for which there is no actively processing WFT).
+  fn eviction_hint(&self, evictable: &[WorkflowSlotInfo]) -> Option<&str>;
```

Comment:
Yeah, the point exactly is to allow them to control the caching "algorithm" without having to actually store the data.

They could pick LRU, or they could pick smallest history first like in @Quinn-With-Two-Ns 's example, or whatever they want.

They can simply return `None` here and we'll use LRU

Reply:
> without having to actually store the data

But that means they have to have access to the cache to even know what's in there (or have their own cache of IDs).

> They could pick LRU

Can they? Do you call this interface every time the cache is _accessed_? Without full control of the caching accesses, one cannot do advanced implementation. I think we should either choose always-LRU or allow custom cache impls. I think it's too difficult to do this hybrid.

Reply:
Ah yeah good point. Seemingly all we'd need to do though is add a `get` or `just_fetched` call, and that'd be sufficient. I don't want it to be a *real* cache implementation because actually storing things won't work well across the language boundary, plus just the intrinsic danger of them returning the wrong item, or losing things that shouldn't be lost, etc. With the facade / sizer even if they screw up we can fall back to something that still operates correctly.

Reply:
Right, if we don't want a real cache impl, let's not even let them choose which thing to evict. Let's keep that complication out.